graphql/api/src/main/resources/stream-registry.graphql

```diff
@@ -33,24 +33,44 @@ type Tag {
 
 ############ Specification ############
 
+input RoleInput {
+    role: String
+    principals: [String]
+}
+
+input SecurityQuery {
+    roleRegex: String
+    principalRegex: String
+}
+
+type Role {
+    role: String
+    principals: [String]
+}
+
+############ Specification ############
+
 input SpecificationInput {
     description: String!
     tags: [TagInput!]!
     type: String!
     configuration: ObjectNode!
+    security: [RoleInput!]!
 }
 
 input SpecificationQuery {
     descriptionRegex: String
     tags: [TagQuery!]
     typeRegex: String
+    security: [SecurityQuery!]
 }
 
 type Specification {
     description: String
     tags: [Tag!]!
     type: String!
     configuration: ObjectNode!
+    security: [Role!]!
```

Comment:
Unfortunately the IT is failing and I think it's because of a mismatch between this being a list and the java in `Specification` being a map.

This is because the return types in the graphql layer are just using the model class and not defining their own types in the graphql layer. This makes it impossible to have a clear separation between the graphql and service layers.

The correct answer is to introduce models into the graphql layer - which would mean mostly a copy & paste except for the difference here with an associated converter. This will greatly grow the scope of the change.

A (short-term?) compromise might be to change the current model to a list (making sure to preserve role uniqueness business logic). This does pollutes the model and introduces tech debt but keeps the scope of the PR smaller.

Reply:
OK - now using List<Role> in the model Specification and GraphQL inputs. Keeping Map style in state models.